Add TwistStamped support via TwistPublisher and TwistSubscriber

Ryanf55 opened the pull request:
## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #1594  |
| Primary OS tested on | Ubuntu 22.04 |
| Robotic platform tested on | CI |

---

## Description of contribution in a few bullet points

* Add a new util class for publishing either Twist or TwistStamped called `TwistPublisher`
* Add a new util class for subscribing to either Twist or TwistStamped called `TwistSubscriber`
* Add a new parameter for selecting to stamp the twist data to each util
* Consume TwistPublisher in nav2_controller, among other nodes
* Consume TwistSubscriber in nav2_velocityy_smoother, among other nodes
* Implements the approach described [here](https://github.com/ros-planning/navigation2/issues/1594#issuecomment-1675262038): `But, I suppose we could make a velocity publisher wrapper in nav2_util that deals with that parameterization outside of the application code (and in a way we can remove the non-stamped support more gradually)`.


## Description of documentation updates required from your changes

* Would need to document the new optional parameter behavior in the normal docs on top of the README.


## Future work that may be required in bullet points

* Related PR's in gazebo, etc, to convert the ecosystem to TwistStamped

## Test Instructions

```bash
colcon test --packages-select nav2_util --event-handlers=console_cohesion+ --ctest-args " -R" " test_twist"
```

Quick iteration
```bash
colcon build --packages-select nav2_util && colcon test --packages-select nav2_util --event-handlers=console_cohesion+ --ctest-args " -R" " test_twist"
```

#### For Maintainers: <!-- DO NOT EDIT OR REMOVE -->
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

Ryanf55 commented on nav2_util/include/nav2_util/twist_publisher.hpp:
```diff
@@ -0,0 +1,128 @@
+public:
+  /**
+  * @brief A constructor
+  * @param service_name name of the service to call
```
copy-pasta errror

Ryanf55 commented on nav2_util/include/nav2_util/twist_publisher.hpp:
```diff
@@ -0,0 +1,128 @@
+    if (is_stamped_) {
+      twist_stamped_pub_->publish(velocity);
+    } else {
+      [[maybe_unused]] const auto cmd_vel = std::make_unique<geometry_msgs::msg::Twist>(
```
remove

SteveMacenski commented on nav2_util/include/nav2_util/twist_publisher.hpp:
```diff
@@ -0,0 +1,127 @@
+
+protected:
+  std::string topic_;
+  nav2_util::LifecycleNode::SharedPtr node_;
```
This doesn't need to be stored

SteveMacenski replied:
There are various reasons why storing the shared pointer creates problems

SteveMacenski commented on nav2_util/include/nav2_util/twist_publisher.hpp:
```diff
@@ -0,0 +1,127 @@
+    }
+  }
+
+  void publish(const geometry_msgs::msg::TwistStamped & velocity)
```
A non-stamped version that adds a stamp if `is_stamped_` is true? Basically the counter-implementation of this one

Ryanf55 replied:
I was thinking about this, and was planning to add it when it's needed, however it seems a lot of NAV2 internals are using the stamped version anyways.

SteveMacenski replied:
Oh ok, then I don't think its strictly necessary then.

The only thing I'd say in counter to that is that if others want to use this to stamp their twists (with a param/method arg for the base_frame) that could be a nice add on. But given your expertise / applications i think you're in the best position to tell me if that's even something desired. If not, feel free to ignore.

SteveMacenski reviewed (commented):
Generally LGTM other than missing test coverage + documentation updates on the new param / feature

Ryanf55 commented on nav2_util/include/nav2_util/twist_publisher.hpp:
```diff
@@ -0,0 +1,127 @@
+  )
+  : topic_(topic), node_(nh)
+  {
+    assert(nh != nullptr);
```
I just removed it, was using it during testing.

SteveMacenski commented:
I think just unit testing is the main blocker here

SteveMacenski commented:
Add unit tests for this new object to make sure its working properly e.g. 
- Check when you ask it to pub with stamp it does
- Check when you ask it to pub with non-stamp it does
- Check the different APIs you've created work / don't crash when called correctly
- Check that get sub count works properly in getting the right stamp vs non-stamp

^ the last 2 are obvious, but we don't know how its going to be modified over time. Also, its dereferencing pointers that will crash if the logic is ever wrong since the other publisher isn't allocated. So exercising the code in CI promises it won't crash applications later. 

Just basic tests - nothing nuts

Ryanf55 commented:
> Add unit tests for this new object to make sure its working properly e.g.
> 
> * Check when you ask it to pub with stamp it does
> * Check when you ask it to pub with non-stamp it does
> * Check the different APIs you've created work / don't crash when called correctly
> * Check that get sub count works properly in getting the right stamp vs non-stamp
> 
> ^ the last 2 are obvious, but we don't know how its going to be modified over time. Also, its dereferencing pointers that will crash if the logic is ever wrong since the other publisher isn't allocated. So exercising the code in CI promises it won't crash applications later.
> 
> Just basic tests with gtest like the other utils - nothing nuts.

Sounds good, can do!

SteveMacenski commented on nav2_util/test/test_twist_publisher.cpp:
```diff
@@ -0,0 +1,53 @@
+// Copyright (c) 2019 Intel Corporation
```
Update

Ryanf55 replied:
Done in 169f7b3778d3198c8266ba7ec026ae5fc66a58bc

SteveMacenski commented on nav2_util/test/test_twist_publisher.cpp:
```diff
@@ -0,0 +1,53 @@
+  vel_publisher->on_activate();
+
+  const geometry_msgs::msg::TwistStamped twist_stamped {};
+  vel_publisher->publish(twist_stamped);
```
This is insufficient. You need to have something that accepts that message to make sure the action is actually taken. Make a subscription and check that it was received / the values. 

Also, Send it both stamped/unstamped and also make sure both are received and the non-stamped was properly stamped. And check that the subscription count is incremented with the subscriber is added. 

Also a second test like this for `enable_stamped_cmd_vel = true` . 

Basically, cover the code :-)

Ryanf55 replied:
Done in 2bc7d9dd2fa0c1a43368927788ad6b2cd1e1f233.

I added a subscriber, and cover both the default value of the param, as well as when the param is modified to enforce stamping. The subscription count is now tested, as well as the activation state and all the transitions.

Ryanf55 commented:
I've added much more comprehensive testing. Currently, CI is failing a test:
https://app.circleci.com/pipelines/github/ros-planning/navigation2/9938/workflows/f5cb1d0d-0c04-48cb-9a51-1a7771b7b613/jobs/31717?invite=true#step-110-9217_47
```
    [ RUN      ] UtilsTests.SmootherTest
    /opt/overlay_ws/src/navigation2/nav2_mppi_controller/test/utils_test.cpp:381: Failure
    Expected: (smoothed_val) < (original_val), actual: 6.3252e+25 vs 11.9772
    [  FAILED  ] UtilsTests.SmootherTest (0 ms)
```

I don't think my PR causes this failure - that test has no reliance on headers I changed. 

As far as all the other uses of Twist in NAV2, I;ve learned some nodes subscribe and republish Twist such as the velocity smoother. In order to have this backward-compatible support for TwistStamped in the entire stack, I expect I'll need to create a mirrored `TwistSubscriber` just like the `TwistPublisher`.

SteveMacenski commented:
I think this PR so that we have a clean migration over to it. Just merging this alone could break users that enable that param and don't have the ability to enable that new param with the velocity smoother and collision monitor. 

I looked over the codecov results and the code and this is good to me now, so consider the publisher approved!

Ryanf55 commented:
I was curious if you had any ideas on the TwistStamped subscribers how to deal with the callbacks needing to have a certain signature. For example: `AssistedTeleop::teleopVelocityCallback`

My thought is to add overrides for all node callbacks to also work with TwistStamped, and convert all internal variables to TwistStamped typed. Then, once the ecosystem (gazebo, etc) is moved over to TwistStamped, the pure Twist callbacks can be deprected on `rolling` and removed in J-turtle.

SteveMacenski commented:
> I was curious if you had any ideas on the TwistStamped subscribers how to deal with the callbacks needing to have a certain signature. For example: AssistedTeleop::teleopVelocityCallback

That sample is pretty easy: the twist subscriber has some `getTwist` method for when it needs to be gotten and otherwise can just be a subobject of this behavior. There may also need to be a `reset()` to set the twist to default (or just have a bool for `received_` that gets reset to mark validness).

Others may be more complex. You may want to enumerate them and we can discuss them individually so you don't waste time

> My thought is to add overrides for all node callbacks to also work with TwistStamped, and convert all internal variables to TwistStamped typed. Then, once the ecosystem (gazebo, etc) is moved over to TwistStamped, the pure Twist callbacks can be deprected on rolling and removed in J-turtle.

I don't fully follow. Like have both a stamped and non-stamped callbacks, each registered, and the stamped just passes the `msg.twist` to the non-stamped version that implements the logic? Or only actually `create_subscription` based on the twist stamped-ness so only one's active on the network

Ryanf55 commented:
> There may also need to be a

Goal: Only subscribe with either Twist or TwistStamped, not both to eliminate overhead.

Based on your feedback, it's leading me to have the TwistSubscriber class constructor take in both signatures (Twist and TwistStamped) of the subscriber callback. It will bind whichever the node is configured for. The problem with that approach, assuming the callbacks are overrides is that the template arg deduction in bind fails. 
```
/home/ryan/Dev/ros2_ws/src/navigation2/nav2_collision_monitor/src/collision_monitor_node.cpp: In member function ‘virtual nav2_util::CallbackReturn nav2_collision_monitor::CollisionMonitor::on_configure(const rclcpp_lifecycle::State&)’:
/home/ryan/Dev/ros2_ws/src/navigation2/nav2_collision_monitor/src/collision_monitor_node.cpp:68:14: error: no matching function for call to ‘bind(<unresolved overloaded function type>, nav2_collision_monitor::CollisionMonitor*, const std::_Placeholder<1>&)’
   68 |     std::bind(&CollisionMonitor::cmdVelInCallback, this, std::placeholders::_1));
      |     ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/11/pstl/glue_algorithm_defs.h:13,
                 from /usr/include/c++/11/algorithm:74,
                 from /opt/ros/rolling/include/rclcpp/rclcpp/executor.hpp:18,
                 from /opt/ros/rolling/include/rclcpp/rclcpp/executors/multi_threaded_executor.hpp:25,
                 from /opt/ros/rolling/include/rclcpp/rclcpp/executors.hpp:21,
                 from /opt/ros/rolling/include/rclcpp/rclcpp/rclcpp.hpp:167,
                 from /home/ryan/Dev/ros2_ws/src/navigation2/nav2_collision_monitor/include/nav2_collision_monitor/collision_monitor_node.hpp:22,
                 from /home/ryan/Dev/ros2_ws/src/navigation2/nav2_collision_monitor/src/collision_monitor_node.cpp:15:
/usr/include/c++/11/functional:789:5: note: candidate: ‘template<class _Func, class ... _BoundArgs> typename std::_Bind_helper<std::__is_socketlike<_Func>::value, _Func, _BoundArgs ...>::type std::bind(_Func&&, _BoundArgs&& ...)’
  789 |     bind(_Func&& __f, _BoundArgs&&... __args)
      |     ^~~~
/usr/include/c++/11/functional:789:5: note:   template argument deduction/substitution failed:
/home/ryan/Dev/ros2_ws/src/navigation2/nav2_collision_monitor/src/collision_monitor_node.cpp:68:14: note:   couldn’t deduce template parameter ‘_Func’
   68 |     std::bind(&CollisionMonitor::cmdVelInCallback, this, std::placeholders::_1));
      |     ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/11/pstl/glue_algorithm_defs.h:13,
                 from /usr/include/c++/11/algorithm:74,
                 from /opt/ros/rolling/include/rclcpp/rclcpp/executor.hpp:18,
                 from /opt/ros/rolling/include/rclcpp/rclcpp/executors/multi_threaded_executor.hpp:25,
                 from /opt/ros/rolling/include/rclcpp/rclcpp/executors.hpp:21,
                 from /opt/ros/rolling/include/rclcpp/rclcpp/rclcpp.hpp:167,
                 from /home/ryan/Dev/ros2_ws/src/navigation2/nav2_collision_monitor/include/nav2_collision_monitor/collision_monitor_node.hpp:22,
                 from /home/ryan/Dev/ros2_ws/src/navigation2/nav2_collision_monitor/src/collision_monitor_node.cpp:15:
/usr/include/c++/11/functional:813:5: note: candidate: ‘template<class _Result, class _Func, class ... _BoundArgs> typename std::_Bindres_helper<_Result, _Func, _BoundArgs>::type std::bind(_Func&&, _BoundArgs&& ...)’
  813 |     bind(_Func&& __f, _BoundArgs&&... __args)
      |     ^~~~
/usr/include/c++/11/functional:813:5: note:   template argument deduction/substitution failed:
/home/ryan/Dev/ros2_ws/src/navigation2/nav2_collision_monitor/src/collision_monitor_node.cpp:68:14: note:   couldn’t deduce template parameter ‘_Result’
   68 |     std::bind(&CollisionMonitor::cmdVelInCallback, this, std::placeholders::_1));
      |     ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
gmake[2]: *** [CMakeFiles/collision_monitor_core.dir/build.make:76: CMakeFiles/collision_monitor_core.dir/src/collision_monitor_node.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:186: CMakeFiles/collision_monitor_core.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2
---
Failed   <<< nav2_collision_monitor [2.51s, exited with code 2]
```

For now, I'll just name the callbacks differently (easy) or switch to lambdas (harder).

Ryanf55 commented:
Hi @SteveMacenski,

I've gotten all the features working. I'd like to start iterating on getting to 100% code coverage, or at least no decrease in code coverage such that you will be happy merging. 

I know that CI runs a tool that shows you a diff of your PR against the target branch, colorizing with respect to changes in code coverage. Is there any way to use it locally so I can iterate faster than waiting for the build and test workflow in CI to run?

Ryanf55 commented:
Looks like the latest build has flake8 failures in CI are totally unrelated to my changes. I do not know what I should do to resolve those:
https://app.circleci.com/pipelines/github/ros-planning/navigation2/10324/workflows/e2990a21-f6d6-428b-8b0c-8029562638b1/jobs/32684?invite=true#step-110-7208_26

SteveMacenski commented on nav2_behaviors/include/nav2_behaviors/plugins/drive_on_heading.hpp:
```diff
@@ -125,23 +126,23 @@ class DriveOnHeading : public TimedBehavior<ActionT>
       return ResultStatus{Status::FAILED, ActionT::Result::COLLISION_AHEAD};
     }
 
-    this->vel_pub_->publish(std::move(cmd_vel));
```
Go back to the unique pointer + move. It removes jitter

SteveMacenski replied:
^ and for all cases like this one

SteveMacenski replied:
I don't know what you're talking about with new syntax. This is Rolling and we have every other publisher using `std::move` on a unique pointer. Please update all publications to use publish moved semantics unique pointers. See pretty much every other publisher in the stack for an example :laughing:

SteveMacenski commented on nav2_behaviors/include/nav2_behaviors/timed_behavior.hpp:
```diff
@@ -289,12 +290,14 @@ class TimedBehavior : public nav2_core::Behavior
+    cmd_vel->twist.linear.y = 0.0;
+    cmd_vel->twist.angular.z = 0.0;
+
+    // vel_pub_->publish(std::move(cmd_vel));
```
Commented otu?

Ryanf55 replied:
Using std::move on the command results in a failure to compile.  It was not clear why TwistStamped needs to be a unique pointer, so I just pass it by value into publish.

Ryanf55 replied:
Here's a reference I found that says the usage I switched to is also valid: 
https://github.com/ros2/rclcpp/pull/676#issuecomment-489381722

What's not clear is why std::move is broken, unless the `rclcpp` maintainers changed the API again since that update.

SteveMacenski commented on nav2_util/include/nav2_util/twist_publisher.hpp:
```diff
@@ -0,0 +1,132 @@
+  {
+    using nav2_util::declare_parameter_if_not_declared;
+    declare_parameter_if_not_declared(
+      node, "enable_stamped_cmd_vel",
```
Config guide needs this for all nodes that use it to know about this option. Migration guide too

SteveMacenski replied:
This is still important before a merge. It should be as easy as adding this parameter to each of the configuration guide pages + the YAML at the bottom -- then writing up a short blurb in the migration guide about it

SteveMacenski commented on nav2_util/include/nav2_util/twist_subscriber.hpp:
```diff
@@ -0,0 +1,92 @@
+        qos,
+        std::forward<TwistCallbackT>(TwistCallback));
+    }
+  }
```
Q: why not have a callback here and a `getTwist` & `getTwistStamped` member?

Ryanf55 replied:
I didn't know what the behavior of getTwistStamped would be if the node received Twist. Does it re-stamp the timestamp automagically? Right now, there is no made up timestamps. Either it has timestamp data, or it doesn't and you aren't allowed to use it.

SteveMacenski replied:
Could throw?

I'm not necessarily asking for a change, just a question to see if it would be sensible to make this similar the `OdomSmoother` or other subscriber-data objects which hold their data and get when requested. I suppose often the Twist Subscription clients are those that need event-based access to the twist data, not polling based, which would be a reasonable answer for why this method is necessary as well.

Ryanf55 replied:
> Could throw?
> 
> I'm not necessarily asking for a change, just a question to see if it would be sensible to make this similar the `OdomSmoother` or other subscriber-data objects which hold their data and get when requested. I suppose often the Twist Subscription clients are those that need event-based access to the twist data, not polling based, which would be a reasonable answer for why this method is necessary as well. So perhaps its just not sensible?

I think it could be sensible, but I worry that it is a bit potential higher impact (probability of bugs). Without sufficient test infrastructure to run the whole system tests with a general parameter override globally for the new param, I'm hesitant to introduce data-sharing modifications.

SteveMacenski reviewed (changes requested):
Overall, LGTM. One thing I'd ask is to review the README files of changed packages + Nav2 documentation to update the use of Twist or TwistStamped. I know its annoying to do docs, but you should be able to get through skimming each of the pages and updating that within 15 minutes and would do a world to help users down the line if they didn't know about this to have hints in the docs easily

SteveMacenski commented:
I fixed the merge conflict

SteveMacenski commented:
Hi,

I think my comments basically were the end of the road of changes needed before merge. If you wanted this in before Thanksgiving, that's definitely on the table if updates are made :-)

Ryanf55 commented:
> Hi,
> 
> I think my comments basically were the end of the road of changes needed before merge. If you wanted this in before Thanksgiving, that's definitely on the table if updates are made :-)

Thanks for the time. I'm swamped this week, but can get those last few things done next week.

SteveMacenski commented:
You should see things pass properly. That looks like just a one-off flaky test. What is your issues locally?

SteveMacenski commented on nav2_velocity_smoother/README.md:
```diff
@@ -59,14 +59,15 @@ velocity_smoother:
    	max_decel: [-2.5, 0.0, -3.2]  # Maximum deceleration, ordered [Ax, Ay, Aw]
    	odom_topic: "odom"  # Topic of odometry to use for estimating current velocities
    	odom_duration: 0.1  # Period of time (s) to sample odometry information in for velocity estimation
+	enable_stamped_cmd_vel: false
```
Add a comment here about what this represents like the others

Ryanf55 replied:
b94181e0b8d6a393d49f3b8e56d8519e8e72c21b

SteveMacenski commented on nav2_util/README.md:
```diff
@@ -7,6 +7,36 @@ The `nav2_util` package contains utilities abstracted from individual packages w
+The behavior up through ROS 2 Iron is preserved - using `Twist`. In a future ROS 2 version, when enough of the
+ROS ecosystem has moved to `TwistStamped`, the default may change. 
+
+### Usage
```
You can remove the usage subsection in entirety

Ryanf55 replied:
Done c9b4e9b271d0afdda09c7c9b307687dfcbb54dc6

SteveMacenski commented on nav2_util/include/nav2_util/twist_publisher.hpp:
```diff
@@ -0,0 +1,132 @@
+    nav2_util::LifecycleNode::SharedPtr node,
+    const std::string & topic,
+    const rclcpp::QoS & qos
+  )
```
`)` on line above, this looks like a strange auto-reformatting choice

Ryanf55 replied:
Interesting. ament_uncrustify accepts code in either configuration as valid. Sorry, I'm used to clang-tidy which always results in the same output regardless of what the input code is.

Fixed!

Ryanf55 replied:
e65fed5e66dac9f0cc264766c078b8b31332f1cd

SteveMacenski commented on nav2_util/include/nav2_util/twist_publisher.hpp:
```diff
@@ -0,0 +1,132 @@
+        qos);
+    } else {
+      twist_pub_ = node->create_publisher<geometry_msgs::msg::Twist>(
+        topic_,
```
Line extension really needed?

Ryanf55 replied:
This is what ament_uncrustify wants. I just reran it.

Ryanf55 replied:
CI is passing on this too.

SteveMacenski commented:
Any update here?

SteveMacenski commented on nav2_velocity_smoother/src/velocity_smoother.cpp:
```diff
@@ -129,9 +129,12 @@ VelocitySmoother::on_configure(const rclcpp_lifecycle::State &)
 
   // Setup inputs / outputs
   smoothed_cmd_pub_ = create_publisher<geometry_msgs::msg::Twist>("cmd_vel_smoothed", 1);
```
This server needs to publish out a TwistStamped, else this PR is essentially not valuable since the pipeline goes from the controller server -> collision monitor -> velocity smoother -> robot base / drone. The outputs of the velocity smoother + collision monitor also need to be twiststamped

SteveMacenski replied:
I think that necessitates the inversion of the changes in this file. The TwistStamped should be the actual callback implementing the actual behavior (e.g. storing header to publish out). The un-stamped Cb should create a header to meet the API to go into the stamped callback to use the rest of the pipeline & publish back out unstamped

Ryanf55 replied:
Done. I also switched it to a unique pointer instead of a shared pointer. 
6b110684405a58ed8a906acd1da7a3aef58af171

SteveMacenski commented on nav2_collision_monitor/src/collision_monitor_node.cpp:
```diff
@@ -63,9 +63,13 @@ CollisionMonitor::on_configure(const rclcpp_lifecycle::State & /*state*/)
-  cmd_vel_in_sub_ = this->create_subscription<geometry_msgs::msg::Twist>(
-    cmd_vel_in_topic, 1,
-    std::bind(&CollisionMonitor::cmdVelInCallback, this, std::placeholders::_1));
+  cmd_vel_in_sub_ = std::make_shared<nav2_util::TwistSubscriber>(
```
Same with velocity smoother: the output here also needs to be stamped and probably changes the way that this should be handled.

Here and in the velocity smoother: why not use the specific constructors for the specific stamped/unstamped parameter so we're not registering both cbs when unnecessary?

Ryanf55 replied:
Yep, good spot. I'll handle both with TwistSubscriber and TwistPublisher. I also modified the internal timestamping for the last command packet to use the incoming packet timestamp rather than node wall clock.

Ryanf55 replied:
50cc29245aafb83b29436de78060caec62170afb

SteveMacenski reviewed (changes requested):
All of the open requests from the last review are also still relevant (but all small). I'll handle the re-pointer-ification once everything else is handled. 

The least of it, but CI is complaining about a small number of linting issues. Usual bits on configuration guide for new parameter + migration guide for capability documentation

SteveMacenski commented:
I handled the conflict

Ryanf55 commented on nav2_util/include/nav2_util/twist_subscriber.hpp:
```diff
@@ -0,0 +1,92 @@
+    nav2_util::LifecycleNode::SharedPtr node,
+    const std::string & topic,
+    const rclcpp::QoS & qos,
+    TwistCallbackT && TwistCallback,
```
I've implemented the support for the `Thought 2` with a second constructor.

Ryanf55 commented on nav2_util/README.md:
```diff
@@ -7,6 +7,36 @@ The `nav2_util` package contains utilities abstracted from individual packages w
+The Twist Publisher and Twist Subscriber are utility classes to assist NAV2 transition from 
+[Twist](https://github.com/ros2/common_interfaces/blob/humble/geometry_msgs/msg/Twist.msg) to [TwistStamped](https://github.com/ros2/common_interfaces/blob/humble/geometry_msgs/msg/TwistStamped.msg).
+
+This utility is in support of the following [GSoC Project](https://navigation.ros.org/2021summerOfCode/projects/twist_n_config.html#convert-twist-to-twiststamped-in-ecosystem-and-run-time-configuration).
```
d074e000c6d6130d0ba6d35b8f19c0af685082ce

Ryanf55 commented:
> All of the open requests from the last review are also still relevant (but all small). I'll handle the re-pointer-ification once everything else is handled.
> 
> The least of it, but CI is complaining about a small number of linting issues. Usual bits on configuration guide for new parameter + migration guide for capability documentation

Thanks. Almost ready. 

Here's the resources on publishing unique_ptr I tried following. Sadly, none of them led me to a resolution with std::move.
https://design.ros2.org/articles/intraprocess_communications.html
![image](https://github.com/ros-planning/navigation2/assets/25047695/b75eca6e-ccb7-4bad-8f73-799e7a6e1179)
https://herbsutter.com/2013/06/05/gotw-91-solution-smart-pointer-parameters/

SteveMacenski commented:
@Ryanf55 I will follow up in a few moments with a PR to supersede this one with some changes. Please review :-)

Ryanf55 commented:
Feel free to re-open this one. I just pushed a fix that @srmainwaring uncovered during testing.
